### PR TITLE
CSHARP-4332: Verify that the examples work in LINQ3 to the extent possible.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+            (c => c.GuidNullable == guidNullable, "{ GuidNullable : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.Guid == invoiceIdNullable, "{ Guid : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.GuidNullable == invoiceIdNullable, "{ GuidNullable : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceId == invoiceIdNullable, "{ InvoiceId : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceIdNullable == invoiceIdNullable, "{ InvoiceIdNullable : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+
+            (c => c.InvoiceId == new InvoiceId(guidNullable.Value), "{ InvoiceId : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceIdNullable == (guidNullable.HasValue ? new InvoiceId(guidNullable.Value) : null), "{ InvoiceIdNullable : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceId == new InvoiceId(guid), "{ InvoiceId : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceIdNullable == new InvoiceId(guid), "{ InvoiceIdNullable : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs:58]: The above 4 test cases were modified to do all type conversions client side.

File: tests/MongoDB.Driver.Tests/MongoDB.Driver.Tests.csproj [hunk 2]
```diff
@@ -6,7 +6,7 @@
   <PropertyGroup>
     <TargetFrameworks>netcoreapp2.1;netcoreapp3.1;net6.0;net472</TargetFrameworks>
     <TargetFrameworks Condition="'$(IsWindows)'!='true'">netcoreapp2.1;netcoreapp3.1;net6.0</TargetFrameworks>
-    <LangVersion>9</LangVersion>
+    <LangVersion>10</LangVersion>
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/MongoDB.Driver.Tests.csproj:9]: Not sure if there is any problem using 10 going forward.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Linq.Linq3ImplementationTests.Jira
+{
+    public class CSharp4332Tests : Linq3IntegrationTest
+    {
+        private static readonly Guid guid = Guid.Parse("0102030405060708090a0b0c0d0e0f10");
+        private static readonly InvoiceId invoiceId = new InvoiceId(guid);
+        private static readonly Guid? guidNullable = (Guid?)guid;
+        private static readonly InvoiceId? invoiceIdNullable = (InvoiceId?) invoiceId;
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs:36]: These should be `_guid`, `_invoiceId`, etc. to follow our naming conventions.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+        // note: when adding or removing test cases be sure to adjust the Range of values for i below
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs:38]: This is awkward. See below for a possible solution.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            (c => c.InvoiceId == new InvoiceId(guidNullable.Value), "{ InvoiceId : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceIdNullable == (guidNullable.HasValue ? new InvoiceId(guidNullable.Value) : null), "{ InvoiceIdNullable : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceId == new InvoiceId(guid), "{ InvoiceId : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+            (c => c.InvoiceIdNullable == new InvoiceId(guid), "{ InvoiceIdNullable : '01020304-0506-0708-090a-0b0c0d0e0f10' }"),
+        };
+
+        [Theory]
+        [ParameterAttributeData]
+        public void Where_expression_should_work(
+            [Range(0, 15)] int i)
```
Comment (JamesKovacs) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp4332Tests.cs:64]: I played around with a few solutions to make this work better. Making the C# compiler happy about types was a bit of a challenge. The following shim works to transform `__testCases` into something that xUnit `MemberDataAttribute` understands. You can likely riff on this idea and come up with something better. The main advantage is that it no longer requires the range to match the array of test cases.

```
        public static IEnumerable<object[]> GetTestCases()
        {
            return __testCases.Select(testCase => new object[] { testCase.Predicate, testCase.ExpectedFilter });
        }

        [Theory]
        [MemberData(nameof(GetTestCases))]
        public void Where_expression_should_work(Expression<Func<Document,bool>> predicate, string expectedFilter)
```

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
Consider using `MemberDataAttribute` instead of `ParameterAttributeDataAttribute`.

**Review by JamesKovacs — APPROVED** [review]
Fair enough point keeping the test case as close to the provided repro as possible. And I see what you mean about xUnit being weird about `Expression`. I had half the tests fail simply by passing the lambda back from `GetTestCases` rather than grabbing it from the array. Your solution balances fixing the fragility of the hard-coded number of test cases (my main concern) and keeping the test cases close to the provided repro. LGTM.

**Pull request closed without merging**